junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/LockManager.java

```diff
@@ -44,18 +48,27 @@ private static Comparator<String> globalKeyFirst() {
 	private final Map<String, ReadWriteLock> locksByKey = new ConcurrentHashMap<>();
 
 	ResourceLock getLockForResources(Collection<ExclusiveResource> resources) {
-		if (resources.size() == 1) {
-			return getLockForResource(getOnlyElement(resources));
-		}
-		List<Lock> locks = getDistinctSortedLocks(resources);
-		return toResourceLock(locks);
+		return toResourceLock(getDistinctSortedLocks(resources));
 	}
 
 	ResourceLock getLockForResource(ExclusiveResource resource) {
-		return new SingleLock(toLock(resource));
+		Lock lock = toLock(resource);
+		if (GLOBAL_READ.equals(resource)) {
+			return new SingleLock.GlobalReadLock(lock);
+		}
+		else if (GLOBAL_READ_WRITE.equals(resource)) {
+			return new SingleLock.GlobalReadWriteLock(lock);
+		}
+		return new SingleLock(lock);
 	}
 
 	private List<Lock> getDistinctSortedLocks(Collection<ExclusiveResource> resources) {
+		if (resources.isEmpty()) {
+			return emptyList();
+		}
+		if (resources.size() == 1) {
+			return singletonList(toLock(getOnlyElement(resources)));
```

Comment:
:x: Isn't this bypassing the globals handling? The old code would invoke `getLockForResource`.
The new code will still use the same underlying `lock` instance but it won't use the correct `SingleLock` subclass.

Reply:
You're right! Done in c87721f9e60035b60ba4354fdcb4cdfd0c1b7739.